### DISCUSSION TITLE
Could you see my kernel if it is wrong?

### DISCUSSION THREAD

**Discussion by bosxered** [idaholab/moose#24961] [Q&A General] [answered]
Hello Moose experts,

I am making a kernel for Qpresidual and QpJacobian so I want to ask Moose experts if my kernel is right.

This is the residual that I want to implement.
![image](https://github.com/idaholab/moose/assets/79071957/f936796a-443d-4561-93fe-9603e1fe2997)

Below is a part of my . C file.

```
Real
NernstElec::computeQpResidual()
{
  return _elec_cond[_qp] * _u[_qp] * _grad_u[_qp] * _grad_test[_i][_qp];
}

Real
NernstElec::computeQpJacobian()
{
  return _elec_cond[_qp] * _grad_test[_i][_qp] * (_phi[_j][_qp] * _grad_u[_qp] + _u[_qp] * _grad_phi[_j][_qp]);
}
```
`_elec_cond[_qp]` is a material property that is considered a constant.
I am not sure, especially, the Jacobian part.

Could you see my kernel if it is wrong?
If it is, can you teach me the right one?
Also, It would be a big help if you tell me the reference from which I can learn how to implement simple Jacobian and also OffDiagJacobian.

Thank you.

Best Wishes,
Jeonghwan

**Comment by GiudGiud** [comment, marked as answer]
Hello

It looks right to me. 
But fortunately you dont have to trust me either, you can use our automated jacobian testing to check:
https://mooseframework.inl.gov/help/development/analyze_jacobian.html

There's no off-diagonal terms here, as there is no other variable than the variable for this equation in this term.

Guillaume

**Reply by bosxered** [reply]
Thank you so much!!
